theano/tensor/sort.py

```diff
@@ -206,3 +218,275 @@ def argsort(a, axis=-1, kind='quicksort', order=None):
+                r = ndim - l
+                reps = list(x.shape)
+                reps[axis] = 1
+                zi = np.arange(abs(k), dtype=idx_dtype)
+                zi = zi.reshape((1,) * l + (k,) + (1,) * (r - 1))
+                zi = np.tile(zi, reps)
+                if op.return_values:
+                    return x.copy(), zi
+                else:
+                    return zi
+
+        idx = [slice(None)] * ndim
+        if k > 0:
+            idx[axis] = slice(-k, None)
+        elif k < 0:
+            idx[axis] = slice(-k)
+        else:
+            raise ValueError('k cannot be zero')
+        if not op.return_indices:
+            zv = np.partition(x, -k, axis=axis)[idx]
+            return zv
+        elif op.return_values:
+            zi = np.argpartition(x, -k, axis=axis)[idx]
+            idx2 = tuple(
+                np.arange(s).reshape(
+                    (s,) + (1,) * (ndim - i - 1)
+                    ) if i != axis else zi for i, s in enumerate(x.shape))
+            zv = x[idx2]
+            return zv, zi.astype(idx_dtype)
+        else:
+            zi = np.argpartition(x, -k, axis=axis)[idx]
+            return zi
+else:
+    def _topk_py_impl(op, x, k, axis, idx_dtype):
+        # TODO better compatibility?
+        raise NotImplementedError('TopKOp: need numpy.argpartition() method (numpy >= 1.8)')
+
+
+class TopKOp(theano.Op):
+    """
+    Operations related to finding k-largest elements.
+
+    The outputs of this Op depends on ``returns_values`` and ``return_indices``,
+    if both ``True``, will return two outputs, corresponding to k-largest values
+    and indices. If only one is ``True``, this Op shall have only one output. Can't
+    be both ``False``.
+
+    Parameters
+    ----------
+    axis: integer
+        The axis to perform the operation. Must be in range ``[-ndim, ndim)``, where
+        ``ndim`` is the dimensionality of input tensor.
+
+    return_values: bool
+        Defaults to ``True``
+
+        If ``True``, one output of the Op will return k-largest array values.
+
+    return_indices: bool
+        Defaults to ``False``
+
+        If ``True``, one output of the Op will return the indices on the given axis.
+
+    Notes
+    -----
+    - ``return_values`` and ``return_indices`` cannot be both ``False``
+
+    See Also
+    --------
+    topk
+    argtopk
+    argtopk_and_topk
+
+    """
+
+    # TODO more params
+    '''
+    sorted: bool
+        Defaults to ``False``
+
+        If True, the result array would be incremental-sorted.
+
+    only_top_kth: bool
+        Defaults to ``False``
+
+        If ``True``, will only find the exact top k-th element. The Op behaves
+        like a reduction.
+    '''
+
+    # TODO c_code
+
+    __props__ = ('axis', 'return_values', 'return_indices', 'idx_dtype')
+
+    def __init__(
+            self,
+            axis=-1,
+            return_indices=False,
+            return_values=True,
+            idx_dtype='int64'):
```

Comment:
I think there are cases where user want to save GPU memory and can ensure smaller types hold it well. There are also [complaints](https://github.com/numpy/numpy/issues/2989) about numpy only using `int64` as arg*() function output dtype.